File: 01-data-structures/03-linked-lists/linked-list_cpp/linked_list.cpp
#include <iostream>

struct Node
{
  Node()
  {
    std::cout << "Node Created!" << std::endl;
  }
  ~Node()
  {
    std::cout << "Node Destroyed!" << std::endl;
  }
  int data;
  Node *next;

};

class List
{
  public:
  List()
  {
    head = NULL;
    tail = NULL;
  }
  void CreateNode(int wholeNumber)
  {
    Node *temp = new Node();
    temp->data = wholeNumber;
    temp->next = NULL;
    
    if (head == NULL)
    {
      head = temp;
      tail = temp;
      temp = NULL;
    }
    else
    {
      tail->next = temp;
      tail = temp;
    }
  }
  
  void DisplayList() 
  {
    Node *temp = new Node();
    temp = head;
    while (temp != NULL)
    {
      std::cout << temp->data << std::endl;
      temp = temp->next;
    }
  }
  
  private:
  Node *head, *tail;
};

int main() 
{
  List list1;
  list1.CreateNode(1);
  list1.CreateNode(3);
  list1.CreateNode(12);
  list1.CreateNode(4);
  list1.DisplayList();


  
  std::cout << "New linked list created" << std::endl;
}

File: 01-data-structures/03-linked-lists/linked_list.rb
require_relative 'node'

class LinkedList
  attr_accessor :head
  attr_accessor :tail

  # This method creates a new `Node` using `data`, and inserts it at the end of the list.
  def add_to_tail(node)

    current = @head
    if (@head ==nil) && (@tail == nil)
      @head = node 
      @tail = node 
    else
      while current.next != nil
        current = current.next
      end
      current.next = node 
      @tail = node
      @tail.next = nil
    end
    
  end

  # This method removes the last node in the lists and must keep the rest of the list intact.
  def remove_tail

    current = @head
    #if no tail do nothing
    if @tail == nil
      return nil
    elsif @head == @tail 
      @head = nil
      @tail = nil
    #else go down the list and remove the tail
    else
      while (current.next != nil) && (current.next != @tail)
        current = current.next
      end
      @tail = current
      @tail.next = nil
    end

  end

  # This method prints out a representation of the list.
  def print
    #print the head
    current = @head
    puts current.data
    #go down the list and print every object
    while current.next != nil
      current = current.next 
      puts current.data
    end

  end

  # This method removes `node` from the list and must keep the rest of the list intact.
  def delete(node)

    current = @head
    if @tail == nil
      return nil
      #list is empty / do nothing
    elsif @head == node 
      @head = @head.next
      #if tail is the node remove the tail
    elsif @tail == node
      remove_tail
    else
      while (current.next != nil) && (current.next != node)
        current = current.next
      end
      current.next = current.next.next
    end

  end

  # This method adds `node` to the front of the list and must set the list's head to `node`.
  def add_to_front(node)

    if (@head == nil) && (@tail == nil)
      @head = node
      @head.next = nil
      @tail = node
      @tail.next = nil
    else
      temp = @head
      @head = node
      @head.next = temp
    end

  end

  # This method removes and returns the first node in the Linked List and must set Linked List's head to the second node.
  def remove_front

    if (@head == nil) || (@head.next == nil) || (@head == @tail)
      @head = nil
    else
      @head = @head.next
    end
  end

end